cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkTuple.java

```diff
@@ -0,0 +1,137 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.spark.data.converter.types.complex;
+
+import java.nio.ByteBuffer;
+import java.util.stream.IntStream;
+
+import org.apache.cassandra.bridge.BigNumberConfig;
+import org.apache.cassandra.spark.data.CqlField;
+import org.apache.cassandra.spark.data.converter.SparkSqlTypeConverter;
+import org.apache.cassandra.spark.data.converter.types.SparkType;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.StructField;
+
+public class SparkTuple implements CollectionFeatures
+{
+    private final SparkSqlTypeConverter converter;
+    private final CqlField.CqlTuple tuple;
+
+    public SparkTuple(SparkSqlTypeConverter converter, CqlField.CqlTuple tuple)
+    {
+        this.converter = converter;
+        this.tuple = tuple;
+    }
+
+    @Override
+    public DataType dataType(BigNumberConfig bigNumberConfig)
+    {
+        return DataTypes.createStructType(IntStream.range(0, size())
+                                                   .mapToObj(index -> DataTypes.createStructField(
+                                                   Integer.toString(index),
+                                                   sparkType(index).dataType(bigNumberConfig),
+                                                   true))
+                                                   .toArray(StructField[]::new)
+        );
+    }
+
+    public CqlField.CqlCollection collection()
+    {
+        return tuple;
+    }
+
+    public SparkSqlTypeConverter converter()
+    {
+        return converter;
+    }
+
+    @Override
+    public Object toSparkSqlType(Object value, boolean isFrozen)
+    {
+        if (value instanceof ByteBuffer)
+        {
+            // Need to deserialize first, e.g. if tuple is frozen inside collections
+            return toSparkSqlType(tuple.deserializeToJavaType((ByteBuffer) value), isFrozen);
+        }
+        else
+        {
+            Object[] array = (Object[]) value;
```

Comment:
nit: How about checking the value type is indeed `Object[]`? And throw IllegalArgumentException, instead of cast error. Maybe there could be some util method to cast types with error message that provides better context on error.